### ISSUE TITLE
Syncing Emails + Comment threads

### ISSUE THREAD

**Issue by CitizensMedia** [the-cm/cm-next#26]
Bob said:     

Logging into CM is a distraction, although it would be better to have many conversations all collected somewhere. Other conversations might be just a distraction for some busy people.

Github, Loomio and google groups (and old school mailing lists) allow all that to happen. In github and Loomio, I often go there to reply to something even though I could do it from email, because it gets back into the email thread, too. So I can find it all in my email history, and everybody can find it in the collection point (whether it be CM or github or Loomio or whatever).

To your point about targeting, yes the email cc's would best be targeted at a specific thread in CM.

**Comment by CitizensMedia** [comment]
Eric said: 
Sounds like there’s an email address for responding to things rather than going to the web interface. Something happens in a thread, people get an email, then they can either:

1. Respond to that email (through email) and something will parse out the interesting part of the email and post it to the thread.
2. Or they can go to the web interface, login, and say what they need to say there.

I have no idea how to set up a mail server to handle this stuff. There are services that can take emails and forward them to a URL (such as https://sendgrid.com/pricing), this way I wouldn’t have to run our own mail server; I don’t know how long it would take to set this up though.

**Comment by CitizensMedia** [comment]
Thanks Eric. If sendgrid offers a sensible solution and it would be the easiest to implement then I say we go for it. Let's wait for Bob to respond then take it from there. We could also reach out to Loomio and ask what they use. Would seem sensible to use the same solution as may help path the way to increased inter-operability.

**Comment by bhaugen** [comment]
Loomio is open source, and our friends. Probably be more willing to help than github, which does the same thing. 

Do you know anybody there to ask? If not, I can probably find somebody.  

Hylo would be another possible willing source of info.

Maybe of interest: here's the reply-to address from a loomio reply-able notification email:
Reply-To: Loomio Community - Feature Ideas <d=135536&u=8381&k=b496d6fe0a797659dd01f24cce869c0d@reply.loomio.org>

Here's one from github:
Reply-To: ssbc/patchwork <reply+0001cabf036d3b403deb921910805ba7bfb1a8fe589f693c92cf0000000112aaab7e92a169ce07601900@reply.github.com>

Here's one from hylo:
Reply-To: <reply-fc8faea97b95b6f739d14e9f907b2342678429accc9cf5a8b2efff59728fca84@mg.hylo.com>

P.S. I'm interested in learning how to do this for NRP, too.

**Comment by bhaugen** [comment]
Loomio is open source, and our friends. Probably be more willing to help than github, which does the same thing. 

Do you know anybody there to ask? If not, I can probably find somebody.  

Hylo would be another possible willing source of info.

Maybe of interest: here's the reply-to address from a loomio reply-able notification email:
```Reply-To: Loomio Community - Feature Ideas <d=135536&u=8381&k=b496d6fe0a797659dd01f24cce869c0d@reply.loomio.org>```

Here's one from github:
```Reply-To: ssbc/patchwork <reply+0001cabf036d3b403deb921910805ba7bfb1a8fe589f693c92cf0000000112aaab7e92a169ce07601900@reply.github.com>```

Here's one from hylo:
```Reply-To: <reply-fc8faea97b95b6f739d14e9f907b2342678429accc9cf5a8b2efff59728fca84@mg.hylo.com>```

P.S. I'm interested in learning how to do this for NRP, too.

**Comment by bhaugen** [comment]
By the way, this is a nice-to-have. I'm not trying to make it urgent for you. You may have many higher priorities.

**Comment by CitizensMedia** [comment]
cool - thanks Bob +1:  you will likely have seen the email to Joshua.. if anyone else comes to mind, please invite them to the discussion. great that it will be of use for NRP too

**Comment by CitizensMedia** [comment]
imperative we get conversations going on CM. the documentary will attract people to the coop which sells itself (and so subsequent investments into the work we are all doing) through our open collaboration around meaningful projects (the pilot being Bergnek). 

therein lies my frustration if you see what I mean. I am trying to help to attract funds through a collaborative story, but people aren't conversing.. unless they do, my vision is a sinking ship.

**Comment by bhaugen** [comment]
Don't be frustrated. Getting engagement in a new conversational medium (new in the sense that it's not ingrained in the habits of the people you want to engage) is hard. Conversations around Bergnek projects should be easier to herd into place once more collaborative work starts to happen. I'll try to engage better myself.

**Comment by mu-is-too-short** [comment]
Loomio and Hylo could be using anything, the email address format is at best only a vague suggestion as to how it works behind the curtains.

Github probably does it internally since they do a lot of volume and have the server and admin resources to manage it.

SendGrid mailhooks work quite simply:

1. A message is sent to `someaddress@x.yourplace.com` for some values of `someaddress`, `x` and `yourplace.com`.
2. Email to `x.yourplace.com` is configured to go to SendGrid's servers rather than your own.
3. SendGrid takes the incoming email, pulls it apart into "to", "from", "subject", and "body" pieces.
4. SendGrid sends those pieces in a standard HTTP request to a URL that you configure at SendGrid.
5. Finally, the URL gets called with those four arguments and whatever you have responding to that URL does whatever it needs to do with the (usually) "to" and "body" chunks. You can see hex-encoded tokens and even full query parameters in the email addresses; any session-like things (such as login tokens to identify when sent the message, which thread/discussion/... the message goes in, ...) will be in the encrypted hex-encoded tokens.

Any other services will work pretty much the same way. Github's approach might be slightly different since they're probably doing it all in-house with their own mail servers but the basic process will be essentially the same everywhere.

**Comment by CitizensMedia** [comment]
thanks Bob.. and you are right. This is why your point about offering email/commenting integration was such a good one! 

thanks Eric.. and sorry this always ends up on your plate

**Comment by bhaugen** [comment]
Actually, @mu-is-too-short - looks to me like you know how to do it. But then it's still (probably picky) work...

**Comment by mu-is-too-short** [comment]
Yes @bhaugen, I've done this sort of thing before. Email clients tend to through a lot of garbage at you though so pulling out just the piece you want is rather finicky work.

**Comment by CitizensMedia** [comment]
Guys at Loomio said:
There is a few layers of technology that enable us to let people reply to emails that appear in the Loomio discussion.

When someone replies to an email that Loomio has sent them, the email goes to [somecode]@reply.loomio.org where [somecode] is a specially crafted address that tells us who the user is and what the discussion is.

As we are an open source application we don't want to depend upon a third party service such as Sendgrid to enable incoming email.

So reply.loomio.org is running a open source NodeJS program called Mailin (http://mailin.io/). It's an inbound SMTP server which sends an API request to ...

A rails app we wrote called Loomio-email-gateway (https://github.com/loomio/loomio-email-gateway). This is a Rails app that reformats the API request so that it works with the Loomio API. It's probably a bit cumbersome to do it this way, if I need to revisit it I'll fold a Mailin compatible API endpoint directly into Loomio.

Finally an API request is sent to Loomio.org and we parse the information about the user, the discussion and the comment body into a new comment. The rest is our normal logic for adding a comment to a thread.. which usually includes sending out more emails.

We've got a gitbook called Loomio Tech Manual, https://loomio.gitbooks.io/tech-manual/content/ which covers some but not all of this. It's in need of an update.
